### PR TITLE
Fix `ModelCheckpoint` with manual optimization and `every_n_train_steps`

### LINKED ISSUE #20947: With automatic_optimization disabled and checkpointing every n steps, the best checkpointed model is the model obtained after backpropagation and not the one used for computing the loss
### Bug description

When the lightning model has disabled the automatic optimization, it is forced to make the optimization process inside the `training_step` method.

In the case of model checkpointing activated with  `every_n_train_steps=1`, the checkpointed model will not be the right one.

Indeed, the model checkpointing is triggered after the method `training_step` (in `on_train_batch_end`), thus the checkpointed model will corresponds to the model on which the backpropagation of the loss has been done.

One workaround is to make the optimization process in  the `on_train_batch_end` of the lightning module and to modify the ModelCheckpoint like following:

```
class CustomModelCheckpoint(ModelCheckpoint):
    """ModelCheckpoint with the ability to save checkpoint when manual optimization is done out of the training step."""
    def __init__(
        self,
        dirpath: None | _PATH = None,
        filename: None | str = None,
        monitor: None | str = None,
        verbose: bool = False,
        save_last: None | Union[bool, Literal["link"]] = None,
        save_top_k: int = 1,
        save_weights_only: bool = False,
        mode: str = "min",
        auto_insert_metric_name: bool = True,
        every_n_train_steps: None | int = None,
        train_time_interval: None | timedelta = None,
        every_n_epochs: None | int = None,
        save_on_train_epoch_end: None | bool = None,
        enable_version_counter: bool = True,
    ):
        super().__init__(
            dirpath=dirpath,
            filename=filename,
            monitor=monitor,
            verbose=verbose,
            save_last=save_last,
            save_top_k=save_top_k,
            save_weights_only=save_weights_only,
            mode=mode,
            auto_insert_metric_name=auto_insert_metric_name,
            every_n_train_steps=every_n_train_steps,
            train_time_interval=train_time_interval,
            every_n_epochs=every_n_epochs,
            save_on_train_epoch_end=save_on_train_epoch_end,
            enable_version_counter=enable_version_counter,
        )

    def _should_skip_saving_checkpoint(self, trainer: Trainer) -> bool:
        from lightning.pytorch.trainer.states import TrainerFn

        return (
            bool(trainer.fast_dev_run)  # type: ignore[attr-defined]
            or trainer.state.fn != TrainerFn.FITTING
            or trainer.sanity_checking
        )
```

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
from copy import deepcopy
from tempfile import TemporaryDirectory

import mlflow
import torch
from lightning import Trainer, LightningModule
from lightning.pytorch.callbacks import ModelCheckpoint
from lightning.pytorch.loggers import MLFlowLogger
from torch.utils.data import Dataset, DataLoader


class FakeDataset(Dataset):
    def __init__(self):
        self.data = [torch.randn(3) for _ in range(4)]
        self.labels = [torch.randint(0, 2, (1,)) for _ in range(4)]

    def __len__(self):
        return 4

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]


def save_model(
    model: torch.nn.Module,
    step_idx: int
) -> None:
    model_copy = deepcopy(model)
    state_dict = model_copy.cpu().state_dict()
    print(f"saving the state dict for {step_idx}")
    torch.save(state_dict, f"model_{step_idx}.pt")


def load_model(
    step_idx: int
) -> None:
    print(f"loading the state dict for {step_idx}")
    return torch.load(f"model_{step_idx}.pt")


class SimpleModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(3, 1)
        self.automatic_optimization = False

        self.fake_losses = [
            torch.tensor(1),
            torch.tensor(1),
            torch.tensor(0),
            torch.tensor(1),
        ]

    def training_step(self, batch, batch_idx):
        out = self.layer(batch[0])
        loss = torch.nn.functional.binary_cross_entropy_with_logits(out, batch[1].float())
        self.log(
            "loss", self.fake_losses[batch_idx], on_step=True, on_epoch=True, logger=True
        )
        save_model(self.layer, batch_idx)

        optimizer = self.optimizers()
        optimizer.zero_grad()
        self.manual_backward(loss)
        optimizer.step()

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)


def create_and_set_experiment(name: str) -> mlflow.entities.Experiment:
    existing_experiment = mlflow.get_experiment_by_name(name)
    if existing_experiment is None:
        mlflow.create_experiment(name)
    return mlflow.set_experiment(experiment_name=name)


mlflow.set_tracking_uri("/storage/ml/mlruns")
mlflow_experiment = create_and_set_experiment(
    name="lightning_checkpoint_issue"
)

with TemporaryDirectory() as tmpdir:
    with mlflow.start_run(experiment_id=mlflow_experiment.experiment_id, run_name="log_issue") as active_run:
        dataset = FakeDataset()
        train_dataloader = DataLoader(dataset, batch_size=1)
        model = SimpleModule()
        trainer = Trainer(
            max_epochs=1,
            accelerator="gpu",
            devices=1,
            logger=MLFlowLogger(
                run_id=active_run.info.run_id,
                tracking_uri="/storage/ml/mlruns",
                log_model=True,
            ),
            callbacks=[
                ModelCheckpoint(
                    save_top_k=1,
                    monitor="loss",
                    dirpath=tmpdir,
                    mode="min",
                    save_last=False,
                    every_n_train_steps=1,
                    train_time_interval=None,
                    every_n_epochs=0,
                    save_on_train_epoch_end=True,
                    save_weights_only=True,
                )
            ],
            log_every_n_steps=1,
            deterministic="warn",
            num_sanity_val_steps=0,
        )
        trainer.fit(model, train_dataloader)

        after_min = load_model(2 + 1) # best loss + 1

        best_ckpt = trainer.checkpoint_callback.best_model_path
        best_ckpt = torch.load(best_ckpt)["state_dict"]

        for layer_name, layer_value in best_ckpt.items():
            assert torch.equal(after_min[layer_name.removeprefix("layer.")],                         layer_value.cpu())
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 4060 Ti
                - NVIDIA GeForce GTX 1080 Ti
        - available:         True
        - version:           12.6
* Lightning:
        - lightning:         2.5.0
        - lightning-utilities: 0.14.3
        - pytorch-lightning: 2.5.1.post0
        - torch:             2.7.1
        - torchmetrics:      1.7.3
* Packages:
        - aiohappyeyeballs:  2.6.1
        - aiohttp:           3.12.13
        - aiosignal:         1.3.2
        - alembic:           1.16.2
        - annotated-types:   0.7.0
        - antlr4-python3-runtime: 4.9.3
        - anyio:             4.9.0
        - async-timeout:     5.0.1
        - attrs:             25.3.0
        - autocommand:       2.2.2
        - backports.tarfile: 1.2.0
        - blinker:           1.9.0
        - cachetools:        5.5.2
        - certifi:           2025.6.15
        - charset-normalizer: 3.4.2
        - click:             8.2.1
        - cloudpickle:       3.1.1
        - contourpy:         1.3.2
        - cycler:            0.12.1
        - databricks-sdk:    0.57.0
        - docker:            7.1.0
        - exceptiongroup:    1.3.0
        - fastapi:           0.115.13
        - filelock:          3.18.0
        - flask:             3.1.1
        - fonttools:         4.58.4
        - frozenlist:        1.7.0
        - fsspec:            2025.5.1
        - gitdb:             4.0.12
        - gitpython:         3.1.44
        - google-auth:       2.40.3
        - graphene:          3.4.3
        - graphql-core:      3.2.6
        - graphql-relay:     3.2.0
        - greenlet:          3.2.3
        - gunicorn:          23.0.0
        - h11:               0.16.0
        - idna:              3.10
        - importlib-metadata: 8.7.0
        - inflect:           7.3.1
        - itsdangerous:      2.2.0
        - jaraco.collections: 5.1.0
        - jaraco.context:    5.3.0
        - jaraco.functools:  4.0.1
        - jaraco.text:       3.12.1
        - jinja2:            3.1.6
        - joblib:            1.5.1
        - kiwisolver:        1.4.8
        - lightning:         2.5.0
        - lightning-utilities: 0.14.3
        - mako:              1.3.10
        - markupsafe:        3.0.2
        - matplotlib:        3.10.3
        - mlflow:            3.1.0
        - mlflow-skinny:     3.1.0
        - more-itertools:    10.3.0
        - mpmath:            1.3.0
        - multidict:         6.5.0
        - networkx:          3.4.2
        - numpy:             2.2.6
        - nvidia-cublas-cu12: 12.6.4.1
        - nvidia-cuda-cupti-cu12: 12.6.80
        - nvidia-cuda-nvrtc-cu12: 12.6.77
        - nvidia-cuda-runtime-cu12: 12.6.77
        - nvidia-cudnn-cu12: 9.5.1.17
        - nvidia-cufft-cu12: 11.3.0.4
        - nvidia-cufile-cu12: 1.11.1.6
        - nvidia-curand-cu12: 10.3.7.77
        - nvidia-cusolver-cu12: 11.7.1.2
        - nvidia-cusparse-cu12: 12.5.4.2
        - nvidia-cusparselt-cu12: 0.6.3
        - nvidia-nccl-cu12:  2.26.2
        - nvidia-nvjitlink-cu12: 12.6.85
        - nvidia-nvtx-cu12:  12.6.77
        - omegaconf:         2.3.0
        - opentelemetry-api: 1.34.1
        - opentelemetry-sdk: 1.34.1
        - opentelemetry-semantic-conventions: 0.55b1
        - packaging:         24.2
        - pandas:            2.3.0
        - pillow:            11.2.1
        - platformdirs:      4.2.2
        - propcache:         0.3.2
        - protobuf:          6.31.1
        - pyarrow:           20.0.0
        - pyasn1:            0.6.1
        - pyasn1-modules:    0.4.2
        - pydantic:          2.11.7
        - pydantic-core:     2.33.2
        - pyparsing:         3.2.3
        - python-dateutil:   2.9.0.post0
        - pytorch-lightning: 2.5.1.post0
        - pytz:              2025.2
        - pyyaml:            6.0.2
        - requests:          2.32.4
        - rsa:               4.9.1
        - scikit-learn:      1.7.0
        - scipy:             1.15.3
        - setuptools:        80.9.0
        - six:               1.17.0
        - smmap:             5.0.2
        - sniffio:           1.3.1
        - sqlalchemy:        2.0.41
        - sqlparse:          0.5.3
        - starlette:         0.46.2
        - sympy:             1.14.0
        - tabulate:          0.9.0
        - threadpoolctl:     3.6.0
        - tomli:             2.2.1
        - torch:             2.7.1
        - torchmetrics:      1.7.3
        - tqdm:              4.67.1
        - triton:            3.3.1
        - typeguard:         4.3.0
        - typing-extensions: 4.14.0
        - typing-inspection: 0.4.1
        - tzdata:            2025.2
        - urllib3:           2.5.0
        - uvicorn:           0.34.3
        - werkzeug:          3.1.3
        - wheel:             0.45.1
        - yarl:              1.20.1
        - zipp:              3.23.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.10.14
        - release:           6.8.0-59-generic
        - version:           #61~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 15 17:03:15 UTC 2

</details>

### More info

_No response_

cc @lantiga

### PR DESCRIPTION
**Pull request by littlebullGit**
## This PR enhances callback to properly handle manual optimization scenarios, ensuring checkpoints reflect the intended model state and providing clear user guidance.

Fixes #20947

### Key Changes:
- **Manual Optimization Support**: Ensures checkpoints capture the model state before optimization when using manual optimization with `every_n_train_steps`.
- **User Warning**: Adds a clear warning when pre-optimization state isn't saved, helping users understand the checkpoint behavior.
- **Documentation**: Updates docstrings and examples to clarify the behavior with manual optimization.

### Testing:
- Added test cases to verify checkpoint behavior with manual optimization
- Ensured backward compatibility with automatic optimization
- Verified warning messages are shown in appropriate scenarios

----
📚 Documentation preview 📚: https://pytorch-lightning--21239.org.readthedocs.build/en/21239/

### CODE REVIEW

**Comment by littlebullGit** [comment]
The link error is (generated/CONTRIBUTING: line    6) broken    https://medium.com/pytorch-lightning/quick-contribution-guide-86d977171b3a - 429 Client Error: Too Many Requests for url: https://medium.com/pytorch-lightning/quick-contribution-guide-86d977171b3a. Not related to my code. The other one is just timed out. 
@Borda @SkafteNicki let me know how to proceed.

**Comment by SkafteNicki** [comment]
> The link error is (generated/CONTRIBUTING: line    6) broken    https://medium.com/pytorch-lightning/quick-contribution-guide-86d977171b3a - 429 Client Error: Too Many Requests for url: https://medium.com/pytorch-lightning/quick-contribution-guide-86d977171b3a. Not related to my code. The other one is just timed out. 
> @Borda @SkafteNicki let me know how to proceed.

Our CI is broken at the moment, nothing you can do. Please stand by while it being fixed.

**Comment by littlebullGit** [comment]
@SkafteNicki @justusschock can we try CI again ? the last run seems fine to me.

File: tests/tests_pytorch/conftest.py [hunk 1]
```diff
@@ -127,9 +128,33 @@ def restore_signal_handlers():
...
     """Ensures that the distributed process group gets closed before the next test runs."""
+    import os
+
+    from lightning.fabric.utilities.port_manager import get_port_manager
+
+    # Record the port used in this test (if any)
+    port_to_release = None
+    if "MASTER_PORT" in os.environ:
+        with contextlib.suppress(ValueError, KeyError):
+            port_to_release = int(os.environ["MASTER_PORT"])
```
Comment (deependujha) [inline tests/tests_pytorch/conftest.py]: shouldn't this code be after `yield` statement?

**Comment by deependujha** [comment]
Hey @littlebullGit, thanks for looking into the port issue! I was digging into the same thing, [mentioned here](https://github.com/Lightning-AI/pytorch-lightning/pull/21195#issuecomment-3416103612).

The changes in this PR don’t fully overlap, so instead of merging directly, it might be cleaner to open a separate PR for this fix if that works for you.

> update - 1:

Borrowed the code to #21195 to see if it works. Thanks for the great work.

> update -2 :

Doesn't seem to work.

File: tests/tests_pytorch/conftest.py (hunk 1 above)
Comment (littlebullGit) [reply tests/tests_pytorch/conftest.py]: You are right. Thank you for pointing out. Both fixtures were checking MASTER_PORT before the test ran (redundant) and then after (the correct place)
I will update now only check after yield to capture the actual port used during the test

**Comment by littlebullGit** [comment]
@deependujha finally I got it fixed. All checks passed for this PR. 
@SkafteNicki @Borda @justusschock My main fix is about Fix ModelCheckpoint. However, the build kept on running into unrelated Address in Use error. So, I took a deeper look and believe fixed that issue as well. Below is my writeup. I cannot find an issue direct related to it. https://github.com/Lightning-AI/pytorch-lightning/issues/2537 seems related, but it is very old. So, let me know if you want me to open a different PR or just fixed in this PR. 

# Port Management Fix for EADDRINUSE Errors

## Problem

Distributed tests were failing in CI with `EADDRINUSE` (address already in use) errors because:

1. **OS TIME_WAIT state**: When tests release ports, the OS keeps them reserved for 30-120 seconds
2. **Immediate reuse**: Test suite immediately tried to reuse the same ports → collision with TIME_WAIT
3. **Untracked spawn ports**: Multi-process tests (DDP spawn) set their own `MASTER_PORT` values that were never tracked
4. **No cleanup mechanism**: Released ports had no protection against immediate reallocation

## Solution: Two-Layer Defense

### Layer 1: Proactive Prevention (Deque)

#### 1. Queue-Based Port Reuse Prevention
- **File**: `src/lightning/fabric/utilities/port_manager.py`
- **What**: `_recently_released` deque with 1024 slots 
- **How**: 
  ```python
  # On release:
  self._recently_released.append(port)
  
  # On allocate:
  if port not in self._recently_released:
      return port  # Safe to use
  ```
- **Why**: Prevents rapid reuse of ports within our process, giving TIME_WAIT sufficient time to clear

#### 2. OS-Based Port Allocation
- **File**: `src/lightning/fabric/utilities/port_manager.py`
- **What**: Use OS's `bind(("", 0))` to pick ports, check against our deque
- **How**:
  ```python
  for attempt in range(max_attempts):
      port = self._find_free_port()  # OS picks port
      
      # Check our tracking structures
      if port not in self._allocated_ports and port not in self._recently_released:
          self._allocated_ports.add(port)
          return port
  ```
- **Why**: OS naturally avoids TIME_WAIT ports (without SO_REUSEADDR), we just prevent intra-process reuse

### Layer 2: Reactive Recovery (Retry Logic)

#### 3. Automatic Retry on EADDRINUSE
- **Files**: `tests/tests_fabric/conftest.py`, `tests/tests_pytorch/conftest.py`
- **What**: Added `pytest_runtest_makereport()` hook that catches EADDRINUSE errors
- **How**:
  ```python
  def pytest_runtest_makereport(item, call):
      if "EADDRINUSE" in str(call.excinfo.value):
          if retry_count < 3:
              manager.release_all()  # Clear state
              time.sleep(0.5)        # Let ports settle
              raise Rerun()          # Retry test
  ```
- **Why**: Handles the rare race window where a port enters TIME_WAIT between our check and TCPStore binding

### Supporting Changes

#### 4. Reserve Externally Assigned Ports
- **File**: `src/lightning/fabric/utilities/port_manager.py`
- **What**: Added `reserve_existing_port()` method
- **Why**: Tracks ports assigned by external launchers (e.g., `torch.distributed.spawn`)

#### 5. Check for Pre-Assigned MASTER_PORT
- **File**: `src/lightning/fabric/plugins/environments/lightning.py`
- **What**: Modified `find_free_network_port()` to check `MASTER_PORT` environment variable
- **Why**: Prevents conflicts when external launchers pre-assign ports

#### 6. Test Fixture Cleanup
- **Files**: `tests/tests_fabric/conftest.py`, `tests/tests_pytorch/conftest.py`
- **What**: Added post-test `MASTER_PORT` capture and cleanup
- **Why**: Tracks ports set by spawned child processes that parent doesn't know about

## Test Coverage

- Added 3 tests for queue-based reuse prevention
- Added 1 test for environment teardown
- Updated 1 test for new allocation approach
- **Total**: 45 tests passing in port_manager suite

## Key Insights from Development

### Why Both Layers Are Needed

**Deque alone isn't enough:**
- If we allocate >1024 ports during TIME_WAIT window (30-120s), we cycle back to ports still in TIME_WAIT
- Example: In heavy CI load, 1024 ports might be exhausted in <30s

**Retry alone isn't enough:**
- Would constantly retry on every rapid reuse scenario
- Wasteful and slow
- Doesn't prevent the problem, just masks it

**Together they work:**
- Deque prevents 99% of conflicts (intra-process reuse)
- Retry handles the 1% edge case (race window)

## Impact

- **✅ Eliminates `EADDRINUSE` errors** in distributed test suites (DDP, FSDP, spawn-based tests)
- **✅ CI now passes** on all port-sensitive tests
- **✅ Addresses root cause** not just symptoms
- **✅ Graceful degradation** via retry logic for unpredictable scenarios
- **✅ No user-facing changes** entirely internal to test infrastructure

## Files Changed

### Core Implementation
- `src/lightning/fabric/utilities/port_manager.py` - Port allocation logic
- `src/lightning/fabric/plugins/environments/lightning.py` - Environment port handling

### Test Infrastructure  
- `tests/tests_fabric/conftest.py` - Fabric test retry logic
- `tests/tests_pytorch/conftest.py` - PyTorch test retry logic
- `tests/tests_fabric/utilities/test_port_manager.py` - Port manager tests

## Commit

**Branch**: `fix/20947-checkpoint-manual-opt`
**Commit**: `5f2079f42`

**Comment by Borda** [comment]
> However, the build kept on running into unrelated Address in Use error. So, I took a deeper look and believe fixed that issue as well. Below is my writeup. I cannot find an issue direct related to it.

Let's move it to its own PR for later tracking or need for further development :flamingo:

**Comment by littlebullGit** [comment]
> EADDRINUSE
https://github.com/Lightning-AI/pytorch-lightning/pull/21309
Please see above PR. I will rebase this PR once that one got merged. Thanks

**Comment by Borda** [comment]
> #21309

merged, pls resolve conflicts :)

**Comment by littlebullGit** [comment]
> > #21309
> 
> merged, pls resolve conflicts :)

@Borda, it has been a while. Any issue with this PR ?
